### ISSUE TITLE
[QUESTION] Ways of running queries

### ISSUE THREAD

**Issue by diegolmello** [Nozbe/WatermelonDB#750]
Hey!
I was reading [this PR](https://github.com/Nozbe/WatermelonDB/pull/747) and I got a question that I don't want to comment there, because it might be just a dumb one...

#### Approach A: query inside of an action
```
await database.action(async () => {
  const post = await postsCollection.find('abcdef')
  await post.update( /* update the post */ )
  await post.markAsDeleted()
})
```

#### Approach B: query outside of an action
```
const post = await postsCollection.find('abcdef')
await database.action(async () => {
  await post.update( /* update the post */ )
  await post.markAsDeleted()
})
```

Is there any difference between Approach A and B?
Maybe because of the way action handle things, it's faster... idk

Thanks!

**Comment by radex** [comment]
@diegolmello actions are an exclusive write lock - all write operations must be contained within an action, and no two actions can execute concurrently. therefore:

- Approach A is safer and more correct - there's no possibility that between `find` and `update`, another piece of code updates or deletes this post
- Approach B, in theory, could be a tiny bit faster, since find can happen even if some action is already running, but in practice, it's unlikely it would make a difference (unless you have some really unusual case, like scheduling 100 actions of the same kind at the same time). It will work perfectly fine 99% of the time, but there is a tiny chance that you run into a race condition, and you'll try to update a deleted post.

In other words: if it's not a problem, go with Approach A -- wrap together logically connected reads and writes into a single action.

PS. The complication is due to Watermelon's asynchronous nature. It gives it a lot of power and flexibility, but sometimes requires users to think about things like this.

**Comment by diegolmello** [comment]
Got it. Thanks! 🙂
